File: ruby/calendar.rb
require 'date'
require 'optparse'

opt = OptionParser.new
option = {}

# 第一週の表示位置を調節する
def first_week_blank(wday)
  #日曜日
  if wday == 0
    print ' ' * 18
  #火曜日~土曜日
  elsif wday >= 2 && wday <= 6
    print '   ' * (wday - 1)
  end
end

# 日曜日には出力後に改行を入れる
def day_judge(day)
  case day.wday
  when 0
    puts sprintf("%2d", day.day)
  when (1..6)
    print sprintf("%2d", day.day) + ' '
  end
end

arg = opt.on('-m month', Integer) {
  |month| option[:month] = month
}
opt.parse!(ARGV)

now = Date.today
year = now.year
select_month = option[:month] || now.month

# 月の範囲外を入力したら、エラーを出す
if select_month < 1 || select_month > 12
  puts "#{option[:month]} is neither a month number (1..12) nor a name"
  exit
end

first_day = Date.new(year, select_month)
first_day_wday = first_day.wday
last_day = Date.new(year, select_month, -1)

puts "#{select_month}月 2023".center(20)
puts "月 火 水 木 金 土 日"
# 第一週の表示位置を合わせる
first_week_blank(first_day_wday)
# 日付を表示
(first_day..last_day).each do |day|
  day_judge(day)
end 
puts ""

File: js/calendar.js
const { program } = require('commander')

program.option('-m, --month, <type>', 'select month')
program.parse(process.argv)

// 土曜日には出力後に改行を入れる
const dayJudge = (day) => {
    if (day.getDay() === 6) {
        console.log((' ' + day.getDate()).slice(-2))
    } else if (day.getDay() >= 0 && day.getDay() <= 5) {
        process.stdout.write((' ' + day.getDate()).slice(-2) + ' ');
    }
}

// 第一週の表示位置を調節する
const firstWeekBlank = (wday) => {
    wdayBlank = wday === 0 ? 7 : wday;
    // process.stdout.write('   '.repeat(wdayBlank - 1));
    process.stdout.write('   '.repeat(wdayBlank));
}

const options = program.opts();
if (Number(options.month) > 12 || Number(options.month) < 1) {
    console.log(`${options.month} is neither a month number (1..12) nor a name`);
    return;
}

const now = new Date();
const selectedMonth = options.month ? Number(options.month) : now.getMonth() + 1;

const firstDay = new Date(2024, selectedMonth - 1, 1);
const firstDayWday = firstDay.getDay();
const lastDay = new Date(2024, selectedMonth, 0);

console.log(`     ${selectedMonth}月 ${now.getFullYear()}`);
console.log('日 月 火 水 木 金 土');
firstWeekBlank(firstDayWday);
for (let day = firstDay; day <= lastDay; day.setDate(day.getDate() + 1)) {
    dayJudge(day);
}
console.log('');